internal/controller/gateway.go

```diff
@@ -30,47 +32,269 @@ import (
-// to watch for Gateway objects across all namespaces.
-func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger) (controller.Controller, error) {
+// to watch for Gateway objects across all namespaces and reconcile those that match class.
+func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger, class string) (controller.Controller, error) {
 	r := &gatewayReconciler{
-		client:       mgr.GetClient(),
-		eventHandler: eventHandler,
-		FieldLogger:  log,
+		ctx:             context.Background(),
+		client:          mgr.GetClient(),
+		eventHandler:    eventHandler,
+		log:             log,
+		classController: class,
 	}
 	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return nil, err
 	}
-	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, &handler.EnqueueRequestForObject{}); err != nil {
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, r.enqueueRequestForOwnedGateway()); err != nil {
 		return nil, err
 	}
 	return c, nil
 }
 
+// enqueueRequestForOwnedGateway returns an event handler that maps events to
+// Gateway objects that reference a GatewayClass owned by Contour.
+func (r *gatewayReconciler) enqueueRequestForOwnedGateway() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		gw, ok := a.(*gatewayapi_v1alpha1.Gateway)
+		if !ok {
+			r.log.WithField("name", a.GetName()).WithField("namespace", a.GetNamespace()).
+				Info("invalid object, bypassing reconciliation.")
+			return []reconcile.Request{}
+		}
+		gc, err := r.classForGateway(gw)
+		if err != nil {
+			r.log.WithField("namespace", gw.Namespace).WithField("name", gw.Name).Error(err)
+			return []reconcile.Request{}
+		}
+		if gc != nil {
+			r.referencedClass = gc
+			// The gateway references a gatewayclass that exists and is managed
+			// by Contour, so enqueue it for reconciliation.
+			r.log.WithField("namespace", gw.Namespace).WithField("name", gw.Name).Info("queueing gateway")
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: gw.Namespace,
+						Name:      gw.Name,
+					},
+				},
+			}
+		}
+		r.log.WithField("name", a.GetName()).WithField("namespace", a.GetNamespace()).
+			Info("gateway not owned by contour, bypassing reconciliation.")
+		return []reconcile.Request{}
+	})
+}
+
+// classForGateway returns an nil, error if the gatewayclass referenced by gw does not exist
+// or nil, nil if gw is not owned by Contour. Otherwise, the referenced gatewayclass is returned.
+func (r *gatewayReconciler) classForGateway(gw *gatewayapi_v1alpha1.Gateway) (*gatewayapi_v1alpha1.GatewayClass, error) {
+	gc := &gatewayapi_v1alpha1.GatewayClass{}
+	if err := r.client.Get(r.ctx, types.NamespacedName{Name: gw.Spec.GatewayClassName}, gc); err != nil {
```

Comment:
>What if there are two classes with the same controller name?

I believe we discussed this in https://github.com/projectcontour/contour/pull/3659. Contour starts up with the `gateway.className` config. Each instance of contour should use a unique `gateway.className`, otherwise they can fight over gatewayclasses with the same `spec.controller`. If we want to set gateway status for this use case, then we need to remove the `classForGateway()` check from the event handler so the gateway is processed and status can be set.